Add Extra ProUI features

classicrocker883 opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description
### This adds the following features to ProUI:  
1. Show Mesh creation / Cancel auto bed leveling, ABL/UBL
- - While having the option to cancel and break mid operation probing, it specifically allows the probe to show the bed map being created for UBL and ABL, `G29`
<br>

2. Ability to change `MULTIPLE_PROBING`/`TOTAL_PROBING` through the **UI** (how many probes when tramming/bed leveling/creating mesh
- - You can variate how many "probes" when probing, such as for bed tramming and creating leveling mesh
<br>

3. Variable Mesh_Inset. Able to change the mesh's inset: 4 sides of Cartesian bed [X/Y] | [Min/Max]
- - You can variate the area of the mesh. Change the **inset** from each edge.
<br>

4. Ability to change Probe Z feedrate (subsequent slow probes) now added
<br>

5. ~~Change the grid array - 3x3, 5x5, 9x9...~~ (removed for future feature)
- - ~~the only limitation I foresee for ProUI is because of the specific MeshViewer, it can only view up to 9x9, but I have another **PR** which allows switching between the two, once that is merged only then we can be able to allow this to go up to 10x10 even 15x15 if need.~~ (goes with changing grid array)
<br>

6. Rearrange menu items
- - Add `levelMenu` + swap w/`advancedSettingsMenu`
- - Keep printer settings in `Advanced Settings` menu and move access to `Control` menu
- - Designate the now `Level` menu for `Bed Leveling` and mesh stuff
- - Preheat menu altered
- - Add `MSG_MESH_SETTINGS` "Mesh Settings"
<br>

## Other Changes
Changes to Gcode Thumbnail preview.  
Move `struct hmiData` from **dwin.h** to **dwin_defines.h**  
Remove redundant headers  
Changed `Probe::run_z_probe` to accommodate variable `MULTIPLE_PROBING` - for ProUI


<br>

>[!NOTE]
> ~~if there is one more thing I'd like to add is being able to change the Z feedrate.~~  
> ~~`MSG_Z_FEED_RATE` goes unused~~ -- *not anymore*
> Having the grid type be 10x10 or more is something we can work on next

~~one more thing is a couple other changes, like the `toggleCheckboxLine()` now uses  `^= true`~~  
- moved to other **PR**
<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->

### Requirements

<!-- Does this PR require a specific board, LCD, etc.? -->

### Benefits
Less clumped.
Better readability.
This goes for UI navigation.
<!-- What does this PR fix or improve? -->

### Configurations

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

### Related Issues

for `void do_z_clearance()`,  
you have `UNUSED(with_probe)`, yet it does get used...  
so I change it to  
`IF_DISABLED(HAS_BED_PROBE, UNUSED(with_probe));`  
I just wanted to point this out, because it just didn't make sense, and hope this change is correct.

~~trying to get the bugs out of `MULTIPLE_PROBING`. its difficult to say which line of code to use. any advice would be helpful.~~
- Multiple probing should be fine, there may be hiccups with HS (High Speed) Mode enabled, but that may depend on bed probe type (3D touch vs BL touch)

---
## Code Below is not relevant to this **PR** anymore since being able to change grid array/grid mesh points was removed.

I get this weird warning only through github actions
```r
Marlin/src/feature/bedlevel/ubl/ubl.cpp: In static member function 'static void unified_bed_leveling::set_all_mesh_points_to_value(const_float_t)':
Marlin/src/feature/bedlevel/ubl/ubl.cpp:107:20: warning: iteration 9 invokes undefined behavior [-Waggressive-loop-optimizations]
  107 |     z_values[x][y] = value;
      |     ~~~~~~~~~~~~~~~^~~~~~~
In file included from Marlin/src/feature/bedlevel/ubl/../../../inc/MarlinConfigPre.h:49,
                 from Marlin/src/feature/bedlevel/ubl/../../../inc/MarlinConfig.h:28,
                 from Marlin/src/feature/bedlevel/ubl/ubl.cpp:23:
Marlin/src/feature/bedlevel/ubl/../../../inc/Conditionals_LCD.h:1568:96: note: within this loop
 1568 |   #define GRID_LOOP(A,B) for (uint8_t A = 0; A < GRID_MAX_POINTS_X; ++A) for (uint8_t B = 0; B < GRID_MAX_POINTS_Y; ++B)
Marlin/src/feature/bedlevel/ubl/ubl.cpp:106:3: note: in expansion of macro 'GRID_LOOP'
  106 |   GRID_LOOP(x, y) {
  ```
and also for `static void unified_bed_leveling::report_current_mesh()`
<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->

thisiskeithb commented:
It would be nice if these kinds of features were implemented at a higher level so everyone could use them instead of a subset of LCDs/UIs.

classicrocker883 reviewed (commented):
> It would be nice if these kinds of features were implemented at a higher level so everyone could use them instead of a subset of LCDs/UIs. 

I'm not so familiar with how that would be; could you point to me which file/line of those, maybe I could figure something out. I'd like to add some more to this.

Crazy-Charles commented on Marlin/src/lcd/e3v2/proui/dwin_defines.h:
```diff
@@ -87,6 +79,45 @@
 #define DEF_HOTENDPIDT TERN(PREHEAT_1_TEMP_BED, PREHEAT_1_TEMP_HOTEND, 195)
 #define DEF_BEDPIDT TERN(PREHEAT_1_TEMP_BED, PREHEAT_1_TEMP_HOTEND, 60)
```
`#define DEF_HOTENDPIDT TERN(PREHEAT_1_TEMP_BED, PREHEAT_1_TEMP_HOTEND, 195)`
`#define DEF_BEDPIDT TERN(PREHEAT_1_TEMP_BED, PREHEAT_1_TEMP_HOTEND, 60)`

Un-related to this PR but these don't work, apart from the obvious typos, PREHEAT_1_TEMP_BED is not used as a boolean, results will always be the int values.

classicrocker883 replied:
you're right, I forgot about this, how it was fixed in my own repo. thanks for the reminder

thinkyhead commented on Marlin/src/feature/bedlevel/abl/bbl.cpp:
```diff
@@ -118,18 +118,18 @@ void LevelingBilinear::set_grid(const xy_pos_t& _grid_spacing, const xy_pos_t& _
   #else
-    constexpr uint8_t ctry1 = (GRID_MAX_CELLS_Y) / 2, // top-of-center
-                      ctry2 = (GRID_MAX_POINTS_Y) / 2,  // bottom-of-center
+    TERN(PROUI_GRID_PNTS, const, constexpr) uint8_t ctry1 = (GRID_MAX_CELLS_Y) / 2,  // top-of-center
```
You can't use `TERN()` with the integer define `PROUI_GRID_PNTS`.

classicrocker883 replied:
wait im confused, `PROUI_GRID_PNTS` is enabled by being defined as 1 - like "true". just as many other examples use `TERN` or `TERN_(HAS_....)`, as in `HAS_whatever` also defined as 1

thinkyhead commented on Marlin/src/core/utility.h:
```diff
@@ -33,7 +33,7 @@ void safe_delay(millis_t ms);           // Delay ensuring that temperatures are
 #endif
 
-#if GRID_MAX_POINTS
+#if TERN(PROUI_GRID_PNTS, HAS_MESH, GRID_MAX_POINTS)
```
You can't use `TERN()` with the integer define `PROUI_GRID_PNTS`. Instead:
```cpp
#if (PROUI_GRID_PNTS > 0 && HAS_MESH) || (PROUI_GRID_PNTS == 0 && GRID_MAX_POINTS > 0)
```

classicrocker883 replied:
I think you may be mixing up that with `hmiData.grid_max_points`, which is the variable integer. `PROUI_GRID_PNTS` simply allows this feature to be used.

I also want to make a note, mostly regarding your recent commit to fix compiling error. so I haven't been able to actually get the variable/customizable `MESH_INSET` or more specifically `MESH_MAX(MIN)_X(Y)` to work properly. in my testing, yes I was able to at least get the values to register and change, but not implement itself when building a mesh. for instance, when editing the values and say setting the mesh to a small square area, it just didnt apply, even though the values were saved and remained after restart.

I just hope I have done it right. you see, I looked through MRiscoC's source code which has the same feature, except some of it is behind the inaccessible library. I might have to completely omit the editable mesh_inset part, because I know the grid points and the other things do work.

any idea what can be done so changing the mesh_inset can be done through the UI?

classicrocker883 replied:
it comes down to the function `get_mesh_x/y()`

so for instance with UBL, this simply did not work:
```y
    static float get_mesh_x(const uint8_t i) {
      return i < (GRID_MAX_POINTS_X) ? pgm_read_float(&_mesh_index_to_xpos[i]) : MESH_MIN_X + i * (MESH_X_DIST);
    }
```

but I got it to work with this (at least for UBL):
```y
    static float get_mesh_x(const uint8_t i) {
      return MESH_MIN_X + i * (MESH_X_DIST);
    }
    static float get_mesh_y(const uint8_t j) {
      return MESH_MIN_Y + j * (MESH_Y_DIST);
    }
```

thinkyhead commented:
I think it would be best to remove the variable mesh grid points part of this PR because we are trying to solve that problem more generally in another PR. The variable points would then apply to all displays and also extend `G29` `M420` `M421` G-codes as needed. This ad hoc solution for ProUI is too "hacky" and invasive to the rest of the code and certainly won't be retained once the general solution is completed.

classicrocker883 reviewed (commented):
> I think it would be best to remove the variable mesh grid points part of this PR because we are trying to solve that problem more generally in another PR. The variable points would then apply to all displays and also extend `G29` `M420` `M421` G-codes as needed. This ad hoc solution for ProUI is too "hacky" and invasive to the rest of the code and certainly won't be retained once the general solution is completed. 

ok yeah I see. can you point to me which PR you're talking about with variable mesh grid points

classicrocker883 commented:
I may need some help figuring out getting Bilinear auto bed leveling to work with the variable mesh inset.
so far UBL works, I haven't yet tested Manual Mesh, but I think a change is needed since it uses the same `get_mesh_x/y`


I might be able to just adjust like so:
like for Manual Mesh bed leveling, add
`#define MESH_X_DIST (float((MESH_MAX_X) - (MESH_MIN_X)) / (GRID_MAX_CELLS_X))`

and in **feature/bedlevel/abl/bbl.h**
```diff
- static float get_mesh_x(const uint8_t i) { return grid_start.x + i * grid_spacing.x; }
- static float get_mesh_y(const uint8_t j) { return grid_start.y + j * grid_spacing.y; }
+ static float get_mesh_x(const uint8_t i) { return MESH_MIN_X + i * (MESH_X_DIST); }
+ static float get_mesh_y(const uint8_t j) { return MESH_MIN_Y + j * (MESH_Y_DIST); }
```
---

but to do it **right**, I think `grid_start` and `grid_spacing` has to be set correctly
fortunately I think I found the code in question for Bilinear:
**Marlin\src\gcode\bedlevel\abl\G29.cpp**
```y
      const float x_min = probe.min_x(), x_max = probe.max_x(),
                  y_min = probe.min_y(), y_max = probe.max_y();

      if (parser.seen('H')) {
        const int16_t size = (int16_t)parser.value_linear_units();
        abl.probe_position_lf.set(_MAX((X_CENTER) - size / 2, x_min), _MAX((Y_CENTER) - size / 2, y_min));
        abl.probe_position_rb.set(_MIN(abl.probe_position_lf.x + size, x_max), _MIN(abl.probe_position_lf.y + size, y_max));
      }
      else {
        abl.probe_position_lf.set(parser.linearval('L', x_min), parser.linearval('F', y_min));
        abl.probe_position_rb.set(parser.linearval('R', x_max), parser.linearval('B', y_max));
      }

...
---
      // Probe at the points of a lattice grid
      abl.gridSpacing.set((abl.probe_position_rb.x - abl.probe_position_lf.x) / (abl.grid_points.x - 1),
                          (abl.probe_position_rb.y - abl.probe_position_lf.y) / (abl.grid_points.y - 1));
```
<br>

as you can see `abl.gridSpacing.set` uses the same format `(MESH_MAX_X) - (MESH_MIN_X)) / (GRID_MAX_CELLS_X)`
and `abl.grid_points.x - 1` = `GRID_MAX_CELLS_X`

for context, `bedleve.set_grid()` sets the values for `grid_spacing`/`grid_start`,  
and is used in `get_mesh_x/y()` 

so where I am stuck is how do I set `probe_position_rb`/`probe_position_lf` to be `MESH_MAX_X`/`MESH_MIN_Y` without it being like a hack job?

thinkyhead committed:
Squash and rebase

thinkyhead committed:
clean up

classicrocker883 committed:
fix compile errors

classicrocker883 commented on Marlin/src/module/motion.h:
```diff
@@ -30,7 +30,7 @@
 #include "../inc/MarlinConfig.h"
 
-#if ALL(DWIN_LCD_PROUI, INDIVIDUAL_AXIS_HOMING_SUBMENU, MESH_BED_LEVELING)
+#if ALL(DWIN_LCD_PROUI, INDIVIDUAL_AXIS_HOMING_SUBMENU, MESH_BED_LEVELING) || (ANY(MESH_BED_LEVELING, AUTO_BED_LEVELING_UBL) && ENABLED(DWIN_LCD_PROUI))
```
@thinkyhead this wouldnt be so wordy if #28231 gets merged

classicrocker883 committed:
tweak to include Bilinear mesh leveling

classicrocker883 committed:
same as before (missed one)

classicrocker883 committed:
more Bilinear condit. -> HAS_MESH + `#ifndef z_probe_slow/fast_mm_s`

classicrocker883 committed:
already under HAS_MESH

classicrocker883 committed:
remove #if defined

classicrocker883 committed:
update probe

classicrocker883 committed:
cosmetic

classicrocker883 committed:
improve drawFillCircle (faster)

classicrocker883 committed:
inline meshviewer

classicrocker883 committed:
removed unused `colorInt`, `drawCricle`

classicrocker883 committed:
relocate

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Jan2

classicrocker883 committed:
fix do_z_clearance with motion.

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Jan2

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26761

thinkyhead committed:
fix some "items"

thinkyhead committed:
extern mesh min/max

thinkyhead committed:
only MBL, UBL

thinkyhead committed:
simplify

classicrocker883 committed:
fix mesh point arrangement

Change variable types from uint8_t to int16_t because it would  offset position when drawing mesh point

classicrocker883 committed:
add to enable PROUI_MESH_EDIT (error fix)

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-Jan2

classicrocker883 committed:
move `dwin_defines.h` to `MarlinConfigPre.h`
- fix parentheses warning
- adjust header file `include`'s

thinkyhead committed:
conditional

thinkyhead committed:
Get ProUI flags in files

thinkyhead committed:
🧑‍💻 Clarify MKS UI flags

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26761

thinkyhead committed:
and others
